File: src/gui/main_window.py

from PyQt6.QtWidgets import (QMainWindow, QTabWidget, QWidget, QVBoxLayout,
                             QHBoxLayout, QPushButton, QLabel, QStatusBar,
                             QMenuBar, QMenu, QToolBar, QDockWidget, QSplitter)
from PyQt6.QtCore import Qt, pyqtSignal
from PyQt6.QtGui import QAction, QIcon

from ai.healing_manager import HealingManager
from core.hotkey_handler import HotkeyHandler


class MainWindow(QMainWindow):
    """Main application window with all development tools"""
    
    def __init__(self, config, device_manager):
        super().__init__()
        self.config = config
        self.device_manager = device_manager
        
        self.setWindowTitle("NiA FBT26 - Flipper Zero Development Suite")
        self.setGeometry(100, 100, 1400, 900)

        # Set up healing manager and hotkey before building UI so that
        # the healing widget can connect to the manager's signals.
        self._healing_manager = HealingManager(
            config=self.config.config if hasattr(self.config, "config") else {},
            parent=self,
        )
        self._hotkey_handler = HotkeyHandler(
            parent_window=self,
            config=self.config.config if hasattr(self.config, "config") else {},
            parent=self,
        )
        self._hotkey_handler.heal_triggered.connect(self._healing_manager.heal)
        
        self.init_ui()

        # Auto-heal on startup if configured
        cfg = self.config.config if hasattr(self.config, "config") else {}
        if cfg.get("healing", {}).get("auto_heal_on_startup", False):
            self._healing_manager.heal()
        
    def init_ui(self):
        # Create menu bar
        self.create_menus()
        
        # Create toolbar
        self.create_toolbar()
        
        # Create central widget with tabs
        self.tabs = QTabWidget()
        self.setCentralWidget(self.tabs)
        
        # Add tabs for each feature
        from .fap_builder import FAPBuilderWidget
        from .firmware_builder import FirmwareBuilderWidget
        from .arduino_panel import ArduinoPanelWidget
        from .terminal_widget import TerminalWidget
        from .github_search import GitHubSearchWidget
        from .healing_widget import HealingWidget
        
        self.tabs.addTab(FAPBuilderWidget(self.config), "FAP Builder")
        self.tabs.addTab(FirmwareBuilderWidget(self.config), "Firmware Builder")
        self.tabs.addTab(ArduinoPanelWidget(self.config), "Arduino/ESP32")
        self.tabs.addTab(TerminalWidget(self.config), "Terminal")
        self.tabs.addTab(GitHubSearchWidget(self.config), "GitHub Search")
        self.tabs.addTab(HealingWidget(self._healing_manager), "AI Orchestration")

        # Connect healing log to status bar
        self._healing_manager.log_message.connect(
            lambda msg: self.statusBar().showMessage(msg, 5000)
        )
        
        # Create status bar
        self.statusBar().showMessage("Ready")
        
    def create_menus(self):
        menubar = self.menuBar()
        
        # File menu
        file_menu = menubar.addMenu("File")
        file_menu.addAction("New Project", self.new_project)
        file_menu.addAction("Open Project", self.open_project)
        file_menu.addSeparator()
        file_menu.addAction("Exit", self.close)
        
        # Device menu
        device_menu = menubar.addMenu("Device")
        device_menu.addAction("Connect", self.connect_device)
        device_menu.addAction("Disconnect", self.disconnect_device)
        device_menu.addSeparator()
        device_menu.addAction("Flash Firmware", self.flash_firmware)
        
        # Tools menu
        tools_menu = menubar.addMenu("Tools")
        tools_menu.addAction("Settings", self.show_settings)
        tools_menu.addSeparator()
        tools_menu.addAction("Heal All AI Services  (Ctrl+Shift+A)",
                             self._healing_manager.heal)
        
    def create_toolbar(self):
        toolbar = self.addToolBar("Main")
        toolbar.addAction("Connect", self.connect_device)
        toolbar.addAction("Build", self.build_project)
        toolbar.addAction("Flash", self.flash_firmware)
        toolbar.addSeparator()
        toolbar.addAction("⚕ Heal", self._healing_manager.heal)
        
    def new_project(self):
        pass
    
    def open_project(self):
        pass
    
    def connect_device(self):
        self.statusBar().showMessage("Connecting to device...")
    
    def disconnect_device(self):
        pass
    
    def flash_firmware(self):
        pass
    
    def build_project(self):
        pass
    
    def show_settings(self):
        pass

File: cpp/src/gui/main_window.cpp
#include "gui/main_window.h"
#include "gui/fap_builder_widget.h"
#include "gui/firmware_builder_widget.h"
#include "gui/arduino_panel_widget.h"
#include "gui/terminal_widget.h"
#include "gui/github_search_widget.h"
#include "core/config_manager.h"
#include "core/device_manager.h"

#include <QAction>
#include <QMessageBox>
#include <QFileDialog>

MainWindow::MainWindow(ConfigManager* config, DeviceManager* deviceManager, QWidget* parent)
    : QMainWindow(parent)
    , m_config(config)
    , m_deviceManager(deviceManager)
    , m_tabs(nullptr)
    , m_fapBuilder(nullptr)
    , m_firmwareBuilder(nullptr)
    , m_arduinoPanel(nullptr)
    , m_terminal(nullptr)
    , m_githubSearch(nullptr)
{
    setWindowTitle("NiA FBT26 - Flipper Zero Development Suite");
    setGeometry(100, 100, 1400, 900);
    
    initUI();
    setupConnections();
}

void MainWindow::initUI()
{
    createMenus();
    createToolbar();
    
    // Create central widget with tabs
    m_tabs = new QTabWidget(this);
    setCentralWidget(m_tabs);
    
    // Create tab widgets
    m_fapBuilder = new FAPBuilderWidget(m_config, this);
    m_firmwareBuilder = new FirmwareBuilderWidget(m_config, this);
    m_arduinoPanel = new ArduinoPanelWidget(m_config, this);
    m_terminal = new TerminalWidget(m_config, this);
    m_githubSearch = new GitHubSearchWidget(m_config, this);
    
    // Add tabs
    m_tabs->addTab(m_fapBuilder, "FAP Builder");
    m_tabs->addTab(m_firmwareBuilder, "Firmware Builder");
    m_tabs->addTab(m_arduinoPanel, "Arduino/ESP32");
    m_tabs->addTab(m_terminal, "Terminal");
    m_tabs->addTab(m_githubSearch, "GitHub Search");
    
    // Create status bar
    statusBar()->showMessage("Ready");
}

void MainWindow::createMenus()
{
    QMenuBar* menubar = this->menuBar();
    
    // File menu
    QMenu* fileMenu = menubar->addMenu("&File");
    fileMenu->addAction("&New Project", this, &MainWindow::newProject, QKeySequence::New);
    fileMenu->addAction("&Open Project", this, &MainWindow::openProject, QKeySequence::Open);
    fileMenu->addSeparator();
    fileMenu->addAction("E&xit", this, &MainWindow::close, QKeySequence::Quit);
    
    // Device menu
    QMenu* deviceMenu = menubar->addMenu("&Device");
    deviceMenu->addAction("&Connect", this, &MainWindow::connectDevice);
    deviceMenu->addAction("&Disconnect", this, &MainWindow::disconnectDevice);
    deviceMenu->addSeparator();
    deviceMenu->addAction("&Flash Firmware", this, &MainWindow::flashFirmware);
    
    // Tools menu
    QMenu* toolsMenu = menubar->addMenu("&Tools");
    toolsMenu->addAction("&Settings", this, &MainWindow::showSettings);
}

void MainWindow::createToolbar()
{
    QToolBar* toolbar = addToolBar("Main");
    toolbar->setMovable(false);
    
    toolbar->addAction("Connect", this, &MainWindow::connectDevice);
    toolbar->addAction("Build", this, &MainWindow::buildProject);
    toolbar->addAction("Flash", this, &MainWindow::flashFirmware);
}

void MainWindow::setupConnections()
{
    connect(m_deviceManager, &DeviceManager::connectionChanged, 
            this, &MainWindow::onConnectionChanged);
    
    connect(m_deviceManager, &DeviceManager::logMessage, 
            m_terminal, &TerminalWidget::appendLog);
    
    connect(m_deviceManager, &DeviceManager::errorOccurred, 
            [this](const QString& error) {
                statusBar()->showMessage("Error: " + error, 5000);
            });
}

void MainWindow::newProject()
{
    statusBar()->showMessage("Creating new project...");
    // Implementation for new project wizard
}

void MainWindow::openProject()
{
    QString dir = QFileDialog::getExistingDirectory(this, 
        "Open Project Directory",
        QDir::homePath(),
        QFileDialog::ShowDirsOnly);
    
    if (!dir.isEmpty()) {
        statusBar()->showMessage("Opened project: " + dir);
    }
}

void MainWindow::connectDevice()
{
    statusBar()->showMessage("Connecting to device...");
    
    if (m_deviceManager->connect()) {
        statusBar()->showMessage("Connected to " + m_deviceManager->deviceName());
    }
}

void MainWindow::disconnectDevice()
{
    m_deviceManager->disconnect();
    statusBar()->showMessage("Disconnected");
}

void MainWindow::flashFirmware()
{
    if (!m_deviceManager->isConnected()) {
        QMessageBox::warning(this, "Warning", "Please connect a device first");
        return;
    }
    
    QString firmware = QFileDialog::getOpenFileName(this,
        "Select Firmware",
        m_config->firmwarePath(),
        "Firmware Files (*.bin *.dfu);;All Files (*)");
    
    if (!firmware.isEmpty()) {
        m_deviceManager->flashFirmware(firmware);
    }
}

void MainWindow::buildProject()
{
    statusBar()->showMessage("Building project...");
    m_fapBuilder->buildFap();
}

void MainWindow::showSettings()
{
    QMessageBox::information(this, "Settings", "Settings dialog coming soon");
}

void MainWindow::onConnectionChanged(bool connected)
{
    if (connected) {
        statusBar()->showMessage("Device connected: " + m_deviceManager->deviceName());
    } else {
        statusBar()->showMessage("Device disconnected");
    }
}
